helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/model/ClusterModelProvider.java

```diff
@@ -200,84 +199,103 @@ private static void retainExistingReplicas(Map<String, Set<AssignableReplica>> r
+          // assignments, there is no need to reassign it.
           String allocatedInstance = commonAllocations.get(0);
           allocatedReplicas.computeIfAbsent(allocatedInstance, key -> new HashSet<>()).add(replica);
-          // Remove the instance from the record to prevent the same location being processed again.
-          baselineAllocations.remove(allocatedInstance);
-          bestPossibleAllocations.remove(allocatedInstance);
-        } else if (!baselineAllocations.isEmpty()) {
-          // 2. If the partition is allocated at an active instance in the Baseline but the
-          // allocation does not exist in the best possible assignment, try to rebalance it.
+          // Remove the instance from the record to prevent this instance from being processed twice.
+          idealAllocations.remove(allocatedInstance);
+          currentAllocations.remove(allocatedInstance);
+        } else if (!idealAllocations.isEmpty()) {
+          // 2. If the partition is allocated at an active instance in the ideal assignment but the
+          // same allocation does not exist in the current assignment, try to rebalance the replica
+          // or assign it if the replica has not been assigned.
+          // There are two possible conditions,
+          // * This replica has been newly added and has not been assigned yet, so it appears in
+          // the ideal assignment and does not appear in the current assignment.
+          // * The allocation of this replica in the ideal assignment has been updated due to a
+          // cluster change. For example, new instance is added. So the old allocation in the
+          // current assignment might be sub-optimal.
+          // In either condition, we add it to toBeAssignedReplicas so that it will get assigned.
           toBeAssignedReplicas.add(replica);
-          // Remove the instance from the baseline record to prevent the same location being picked
-          // up again.
-          baselineAllocations.remove(baselineAllocations.iterator().next());
-        } else if (!bestPossibleAllocations.isEmpty()) {
-          // 3. If the partition is allocated at an active instance in the best possible assignment
-          // only, there is no need to rebalance it.
-          String allocatedInstance = bestPossibleAllocations.iterator().next();
+          // Remove the pending allocation from the idealAllocations after processing so that the
+          // instance won't be double-processed as we loop through all replicas
+          String pendingAllocation = idealAllocations.iterator().next();
+          idealAllocations.remove(pendingAllocation);
+        } else if (!currentAllocations.isEmpty()) {
+          // 3. This replica exists in the current assignment but does not appear or has a valid
+          // allocation in the ideal assignment.
+          // This means the ideal assignment contains an allocation that the instance is temporary
```

Comment:
Rephrase to make it clear:

          // This means either 1) that the ideal assignment actually has this replica allocated on this instance, but it does not show up because the instance is temporarily offline or disabled (note that all such instances have been filtered out in earlier part of the logic) or that the most recent version of the ideal assignment was not fetched correctly from the assignment metadata store.

Reply:
changed